### PR TITLE
Fx Trampoline Coroutine Context

### PR DESCRIPTION
**Pull request by raulraja**
Fixes stack safety on `!effect { foo() }` jumping async after 127 binds.
To be improved and configurable.

### CODE REVIEW

File: modules/effects/arrow-effects-io-extensions/src/main/kotlin/arrow/effects/IODispatchers.kt [hunk 1]
```diff
@@ -55,7 +61,12 @@ object IODispatchers {
     override fun resumeWith(result: Result<T>) {
       if (iterations.get() > asyncBoundaryAfter) {
         iterations.set(0)
-        pool.execute { cont.resumeWith(result) }
+        println("Jumping async. cont.context: $context")
```
Comment (pakoito) [inline modules/effects/arrow-effects-io-extensions/src/main/kotlin/arrow/effects/IODispatchers.kt]: Remove?

File: modules/effects/arrow-effects-io-extensions/src/main/kotlin/arrow/effects/internal/Trampoline.kt [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      underlying.execute(ResumeRun(head, rest))
+    }
+  }
+
+  private tailrec fun immediateLoop(task: Runnable): Unit {
+    try {
+      task.run()
+    } catch (ex: Throwable) {
+      forkTheRest()
+      throw ex.nonFatalOrThrow()
```
Comment (pakoito) [inline modules/effects/arrow-effects-io-extensions/src/main/kotlin/arrow/effects/internal/Trampoline.kt]: Should we throw this? I thought it throwed internally

File: modules/effects/arrow-effects-io-extensions/build.gradle [hunk 3]
```diff
@@ -1,6 +1,7 @@
 apply plugin: 'kotlin-kapt'
 
 dependencies {
+    compile project(':arrow-free-data')
```
Comment (pakoito) [inline modules/effects/arrow-effects-io-extensions/build.gradle]: Free isn't used, it should not be necessary.

**Comment by raulraja** [comment]
We are still working in the design in #1299 for a 0.9.1 solution in a different branch that is more performant than this one.

**Comment by raulraja** [comment]
This is part of IO now in master

**Pull request closed without merging**